filecoin-proofs/src/api/post_adapter.rs

```diff
@@ -0,0 +1,626 @@
+    pub sector_bytes: PaddedBytesAmount,
+    pub comm_rs: [Commitment; POST_SECTORS_COUNT],
+    pub challenge_seed: ChallengeSeed,
+    pub proof: [u8; 192],
+    pub faults: Vec<u64>,
+}
+
+pub struct GeneratePoStDynamicSectorsCountOutput {
+    pub proofs: Vec<[u8; 192]>,
+    pub faults: Vec<u64>,
+}
+
+pub struct GeneratePoStFixedSectorsCountOutput {
+    pub proof: [u8; 192],
+    pub faults: Vec<u64>,
+}
+
+pub struct VerifyPoStDynamicSectorsCountOutput {
+    pub is_valid: bool,
+}
+
+pub struct VerifyPoStFixedSectorsCountOutput {
+    pub is_valid: bool,
+}
+
+/// Maps inputs for a single dynamic sectors-count PoSt generation operation to
+/// a vector of PoSt generation inputs for fixed sectors-count.
+///
+pub fn generate_post_spread_input(
+    dynamic: GeneratePoStDynamicSectorsCountInput,
+) -> Vec<GeneratePoStFixedSectorsCountInput> {
+    let mut fixed: Vec<GeneratePoStFixedSectorsCountInput> = vec![];
+
+    let chunks = dynamic.input_parts.chunks_exact(POST_SECTORS_COUNT);
+
+    let remainder = chunks.remainder();
+
+    for chunk in chunks {
+        fixed.push(GeneratePoStFixedSectorsCountInput {
+            sector_bytes: dynamic.sector_bytes,
+            challenge_seed: dynamic.challenge_seed,
+            input_parts: [chunk[0].clone(), chunk[1].clone()],
+        });
+    }
+
+    if !remainder.is_empty() {
+        let mut input_parts: [(Option<String>, Commitment); POST_SECTORS_COUNT] =
+            Default::default();
+
+        let vs = repeat_with(|| remainder[remainder.len() - 1].clone());
+        let rs = remainder.iter().cloned();
+        let it = rs.chain(vs).take(POST_SECTORS_COUNT).enumerate();
+
+        for (i, input_part) in it {
+            input_parts[i] = input_part;
+        }
+
+        fixed.push(GeneratePoStFixedSectorsCountInput {
+            sector_bytes: dynamic.sector_bytes,
+            challenge_seed: dynamic.challenge_seed,
+            input_parts,
+        });
+    }
+
+    fixed
+}
+
+/// Collapses return values from multiple fixed sectors-count PoSt generation
+/// operations into a single return value for dynamic sector count. This
+/// function discards any faults reported for duplicated CommRs.
+///
+pub fn generate_post_collect_output(
+    orig_comm_rs_len: usize,
+    xs: Vec<error::Result<GeneratePoStFixedSectorsCountOutput>>,
+) -> error::Result<GeneratePoStDynamicSectorsCountOutput> {
+    if xs.is_empty() {
+        return Err(format_err!("input vector must not be empty"));
+    }
+
+    let mut dynamic = GeneratePoStDynamicSectorsCountOutput {
+        proofs: Vec::new(),
+        faults: Vec::new(),
+    };
+
+    for (i, x) in xs.into_iter().enumerate() {
+        match x {
+            Err(e) => {
+                return Err(e);
+            }
+            Ok(fixed) => {
+                dynamic.proofs.push(fixed.proof);
+                for fault in fixed.faults.iter() {
+                    dynamic.faults.push(((i as u64) * 2) + fault)
```

Comment:
Not sure what's going on here, where did the `2` come from?

Reply:
That's a bug. The `2` should be `POST_SECTORS_COUNT`.

> Not sure what's going on here

This awkward code maps values in the fixed-output `faults` vectors to their equivalents in the dynamic-outputs `faults` vectors. 

The formula is this: `dynamic fault value = static fault value + (static output index * POST_SECTORS_COUNT`.

I dunno if this helps, but here's an attempt at a digram:

```
static_outputs = [
  GeneratePoStFixedSectorsCountOutput(faults=[]), 
  GeneratePoStFixedSectorsCountOutput(faults=[0]),
  GeneratePoStFixedSectorsCountOutput(faults=[1])
]

dynamic_outputs = generate_post_collect_output(static_outputs)
// GeneratePoStDynamicSectorsCountOutput(faults=[2, 5])
```